docs/design/Ray-Job-Manager.md

```diff
@@ -0,0 +1,78 @@
+# Ray TPU Job Manager
+
+This is a quick design document to explain how our Ray TPU Job Manager works.
+
+## Introduction
+
+Please see the [Ray documentation](https://docs.ray.io/en/latest/index.html) for more information.
+Ray is a resource-aware job scheduler, so you can specify the resources that a job requires:
+
+```python
+@ray.remote(num_cpus=4)
+def my_cpu_job():
+    ...
+```
+
+For GPUs, Ray lets you specify the number of GPUs you need:
+
+```python
+@ray.remote(num_gpus=1)
+def my_gpu_job():
+    ...
+```
+
+In Ray, TPUs are roughly represented the same way, but there are a number of problems with that approach.
+In particular:
+
+* Ray's granularity allows it to schedule a task on a single machine, not across multiple machines. In particular,
+Ray can't directly schedule a task on a TPU slice that spans multiple machines.
+* Google requires that only one process on a machine can access the TPU at a time.
+
+This document explains how we work around those problems.
+
+### Scheduling Slices of TPUs
+
+TPU slices must be used in a SPMD manner (this is probably not quite true, but it's the easiest way to use them).
+This means that you need to schedule the same task on multiple slices of the TPU at once. Ray can't do this directly.
+
+Allen Wang (@allenwang28) at Google wrote [this gist](https://gist.github.com/allenwang28/e3400b9e9212b50aa1cda55ebeccea60#file-ray_tpu_task-py)
+that is most of the way to a solution. The key idea is to schedule a task on the special `"TPU-XXXX-head"` resource
+(where XXXX is the TPU type like `"v4-256"`). This gives you a "lock" on the TPU slice. Once you have the lock, you can
+query the VM to get a unique resource that is shared only for this particular slice. You can then use this resource to
+schedule K tasks on the K workers that are on the same slice. These tasks do the actual work.
+
+Managing preemption then is just a question of rescheduling the job when it gets preempted: getting a new lock and
+restarting the job.
+
+### Dealing with `libtpu`
+
+`libtpu` is a library that interfaces with the TPU. `libtpu` has a hard requirement that only one process on a machine
+can access the TPU at a time. It manages this with a lockfile called `/tmp/libtpu_lockfile`. Ordinarily, this is fine,
+as the lockfile is removed when the process exits. However, Ray maintains a pool of workers that don't ordinarily exit
+between tasks. This means that the lockfile is not removed, and the next job that tries to access the TPU will fail.
+
+As best I can tell, it's actually fine to remove this lockfile so long as you're not trying to access the TPU from
```

Comment:
IIRC the kernel driver file can only be opened by a single process exclusively, but I think might have been added to help reduce confusion if 2 processes started almost at the same time?

In any case, as long as only one process at a time is using the TPU it's fine.

Reply:
gotcha. Well, I'm just gonna nuke the lockfile and hope for the best.